virttest/utils_net.py

```diff
@@ -3744,3 +3746,195 @@ def map_hostname_ipaddress(hostname_ip_dict, session=None):
+    pdb_path = session.cmd(pdb_find_cmd).strip()
+    logging.info("Found %s.pdb file at %s" % (driver_name, pdb_path))
+    return pdb_path
+
+
+def _prepare_traceview_windows(test, params, vm, session, timeout=360):
+    """
+    Copy traceview.exe and corresponding pdb file to drive c: for future use.
+
+    :param test: the running test
+    :param params: the test params
+    :param vm: target vm
+    :param session: a session to send command
+    :param timeout: the command execute timeout
+    :return: a tuple which consists of local traceview.exe and pdb file paths
+    """
+
+    copy_cmd = "xcopy %s %s /y"
+    dst_folder = "c:\\"
+    # copy traceview.exe
+    logging.info("Copy traceview.exe to drive %s" % dst_folder)
+    traceview_path = _get_traceview_path(session, params)
+    session.cmd(copy_cmd % (traceview_path, dst_folder))
+
+    # copy Netkvm.pdb
+    driver_name = params.get("driver_name", "netkvm")
+    logging.info("Locate %s.pdb and copy to drive %s" %
+                 (driver_name, dst_folder))
+    pdb_path = _get_pdb_path(session, driver_name)
+    session.cmd(copy_cmd % (pdb_path, dst_folder))
+
+    # return local file names
+    pdb_local_path = "%s%s.pdb" % (dst_folder, driver_name)
+    traceview_local_path = dst_folder + "traceview.exe"
+    return (pdb_local_path, traceview_local_path)
+
+
+def _get_MSI_queue_from_traceview_output(output):
+    """
+    Extract msi&queues infomation from traceview log file output
+
+    :param output: the content of traceview processed log infomation
+    :return: a tuple of (msis, queues)
+    """
+    info_str = "Start checking dump content for msi & queues info"
+    logging.info(info_str)
+    search_exp = r'No MSIX, using (\d+) queue'
+    # special case for vectors = 0
+    queue_when_no_msi = re.search(search_exp, output)
+    if queue_when_no_msi:
+        return (0, int(queue_when_no_msi.group(1)))
+    search_exp = r'(\d+) MSIs, (\d+) queues'
+    search_res = re.search(search_exp, output)
+    if not search_res:
+        return (None, None)
+
+    MSIs_number = int(search_res.group(1))
+    queues_number = int(search_res.group(2))
+    return (MSIs_number, queues_number)
+
+
+def _wait_for_traceview_dump_finished(session, dump_file_path, timeout=100):
+    """
+    Check the dump file size periodically, untill the file size doesn't change,
+    considered the dump process has finished. Then kill the idled progress.
+
+    :param session: a session to send command
+    :param dump_file_path: the dump file to check
+    """
+    def _get_file_size(last_size, file_path):
+        check_file_size_cmd = "for %%I in (%s) do @echo %%~zI" % file_path
+        status, output = session.cmd_status_output(check_file_size_cmd)
+        if status or not output.isdigit():
+            return False
+        file_size = int(output)
+        if file_size != last_size[0]:
+            last_size[0] = file_size
+            return False
+        return True
+
+    last_size = [0]
```

Comment:
Why need a list for it? Seems you only need one element.

Reply:
The purpose to use list is that we need to record the newest file size each time the 

>` _get_file_size`

 is called

Reply:
The purpose to use list is that we need to record the newest file size each time the `_get_file_size` is called. The use the newest value for the next call to this function. If only a variable is used as parameter, it's only passed as value and can't be changed inside the function. So we use a list to store the new size, as it's passed by reference and its element can be changed and restored. Other suggestion has been solved :), pls help to review again

Reply:
@kimi1978wy I don't really understand, `_get_file_size()` is defined under `_wait_for_traceview_dump_finished()`, it's in the same context with `last_size`, you don't even need to pass `last_size` to the function, you might directly use a 'global' `last_size`.  Also for `file_path`,  the `check_file_size_cmd` does not need to be generated each time calling the function. 
```
def _get_file_size():
        status, output = session.cmd_status_output(check_file_size_cmd)
        if status or not output.isdigit():
            return False
        file_size = int(output)
        if file_size != last_size:
            last_size = file_size
            return False
        return True

check_file_size_cmd = "for %%I in (%s) do @echo %%~zI" % dump_file_path
last_size = 0
utils_misc.wait_for(lambda: _get_file_size()...)
```

Reply:
And the function name `_get_file_size` does not match the implementation very well, maybe something like `_check_file_size_stable` would be better, also better have at least one line docstring for it.

Reply:
@vivianQizhu  Updated, pls help to review  it again, thanks